net/xia/ppal_ether/main.c

```diff
@@ -168,6 +168,263 @@ static void local_free_ether(struct fib_xid_table *xtbl, struct fib_xid *fxid)
 	kfree(leid);
 }
 
+/* Main-FXID. */
+struct fib_xid_ether_main {
+	struct xip_dst_anchor   xem_anchor;
+	struct ether_interface  *host_interface;
+	struct list_head        interface_common_addr;
+
+	/* WARNING: @xhm_common is of variable size, and
+	 * MUST be the last member of the struct.
+	 */
+	struct fib_xid          xem_common;
+};
+
+/* Main-FXID functions. */
+static inline struct fib_xid_ether_main *fxid_mether(struct fib_xid *fxid)
+{
+	return likely(fxid)
+		? container_of(fxid, struct fib_xid_ether_main, xem_common)
+		: NULL;
+}
+
+/* Main-FXID + INTERFACE. */
+static void attach_neigh_to_interface(struct fib_xid_ether_main *mether)
+{
+	ether_interface_hold(mether->host_interface);
+	/* When using list_add_tail_rcu the caller must take whatever
+	 * precautions are necessary (such as holding appropriate locks) to
+	 * avoid racing with another list-mutation primitive, such as
+	 * list_add_tail_rcu() or list_del_rcu(), running on this same list.
+	 */
+	spin_lock(&mether->host_interface->interface_lock);
+	list_add_tail_rcu(&mether->host_interface->list_interface_common_addr,
+			  &mether->interface_common_addr);
+	spin_unlock(&mether->host_interface->interface_lock);
+	/* Don't call ether_interface_put(einterface) here because @mether
+	 * is in its list.
+	 */
+}
+
+/* Main-FXID - INTERFACE. */
+static void detach_neigh_to_interface(struct fib_xid_ether_main *mether)
+{
+	/* When using list_del_rcu the caller must take whatever precautions
+	 * are necessary (such as holding appropriate locks) to avoid racing
+	 * with another list-mutation primitive, such as list_add_tail_rcu()
+	 * or list_del_rcu(), running on this same list.
+	 */
+	spin_lock(&mether->host_interface->interface_lock);
+	list_del_rcu(&mether->interface_common_addr);
+	spin_unlock(&mether->host_interface->interface_lock);
+
+	ether_interface_put(mether->host_interface);
+}
+
+/* ETHER main table operations. */
+static int main_newroute(struct xip_ppal_ctx *ctx, struct fib_xid_table *xtbl,
+			 struct xia_fib_config *cfg)
+{
+	struct xip_deferred_negdep_flush *dnf;
+	struct fib_xid_ether_main *mether;
+	struct net_device *out_interface;
+	struct fib_xid *cur_fxid;
+	u32 nl_flags, bucket;
+	u32 *p;
+	const char *id;
+	int rc, lladdr_len;
+	u8 lladdr[MAX_ADDR_LEN];
+
+	/* Check for errors in cfg. */
+	if (!cfg->xfc_dst)
+		return -EINVAL;
+
+	p = (u32 *)cfg->xfc_dst->xid_id;
+
+	ASSERT_RTNL();
+	out_interface = __dev_get_by_index(ctx_ether(ctx)->net,
+					   __be32_to_cpu(*p));
+	if (!out_interface)
+		return -ENODEV;
+
+	BUILD_BUG_ON((XIA_XID_MAX - sizeof(*p)) > MAX_ADDR_LEN);
```

Comment:
The current code compiles correctly as u can see that the bug should be built when the address length in our paramters is greater than the maximum allowed link layer address length. If the sign is changed it crashes.

Reply:
The reason it compiles is that `MAX_ADDR_LEN` is way too big (i.e. 32). The test must be `<`. Replace `BUILD_BUG_ON` with `BUG_ON`, replace `MAX_ADDR_LEN` with `out_interface->addr_len`, and drop the following line, namely, `BUG_ON(out_interface->addr_len > MAX_ADDR_LEN);`. This will solve the issue. Thank you for testing it.